packages/grafana-flamegraph/src/FlameGraph/rendering.ts

```diff
@@ -60,7 +66,130 @@ export function useFlameRender(options: RenderOptions) {
     colorScheme,
     focusedItemData,
   } = options;
-  const foundLabels = useMemo(() => {
+  const foundLabels = useFoundLabels(search, data);
+  const ctx = useSetupCanvas(canvasRef, wrapperWidth, depth);
+  const theme = useTheme2();
+
+  // There is a bit of dependency injections here that does not add readability, mainly to prevent recomputing some
+  // common stuff for all the nodes in the graph when only once is enough. perf/readability tradeoff.
+
+  const getBarColor = useColorFunction(
+    totalColorTicks,
+    totalTicksRight,
+    colorScheme,
+    theme,
+    rangeMin,
+    rangeMax,
+    foundLabels,
+    focusedItemData ? focusedItemData.item.level : 0
+  );
+  const renderFunc = useRenderFunc(ctx, data, getBarColor, textAlign);
+
+  useEffect(() => {
+    if (!ctx) {
+      return;
+    }
+
+    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);
+    walkTree(root, direction, data, totalViewTicks, rangeMin, rangeMax, wrapperWidth, renderFunc);
+  }, [ctx, data, root, wrapperWidth, rangeMin, rangeMax, totalViewTicks, direction, renderFunc]);
+}
+
+type RenderFunc = (
+  item: LevelItem,
+  x: number,
+  y: number,
+  width: number,
+  height: number,
+  label: string,
+  // Collapsed means the width is too small to show the label, and we group collapsed siblings together.
+  collapsed: boolean
+) => void;
+
+function useRenderFunc(
```

Comment:
Separated lots of the stuff to smaller functions mainly for better memoization between renders or for easier testing.